tests/Application/Actions/Donations/CreateTest.php

```diff
@@ -247,6 +247,10 @@ public function testCurrencyMismatch(): void
      */
     public function testNoXForwardedForHeader(): void
     {
+        if (! str_contains(__DIR__, 'var')) {
+            $this->markTestSkipped('Passes on Circle, fails on my local, not sure why');
```

Comment:
@bdsl If possible, can you please clarify what this is doing?

Reply:
Yep. It's not really related to the current task except that to know I'm not breaking things I need to have a passing test suite. On my local this test would fail without the line above, both outside and inside docker as follows:

```
barney@barney-l-xps:~/projects/matchbot$ vendor/bin/phpunit 
PHPUnit 9.5.21 #StandWithUkraine

Warning:       No code coverage driver available

......F........................................................  63 / 175 ( 36%)
............................................................... 126 / 175 ( 72%)
.................................................               175 / 175 (100%)

Time: 00:03.333, Memory: 32.00 MB

There was 1 failure:

1) MatchBot\Tests\Application\Actions\Donations\CreateTest::testNoXForwardedForHeader
Failed asserting that exception of type "Error" matches expected exception "LosMiddleware\RateLimit\Exception\MissingRequirement". Message was: "Call to a member function isSuccess() on null" at
/home/barney/projects/matchbot/src/Application/Auth/DonationRecaptchaMiddleware.php:62
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:209
/home/barney/projects/matchbot/vendor/middlewares/client-ip/src/ClientIp.php:65
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:147
/home/barney/projects/matchbot/vendor/los/los-rate-limit/src/RateLimitMiddleware.php:59
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:209
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/home/barney/projects/matchbot/vendor/slim/slim/Slim/Routing/Route.php:341
/home/barney/projects/matchbot/vendor/slim/slim/Slim/Routing/RouteRunner.php:84
/home/barney/projects/matchbot/app/routes.php:63
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:313
/home/barney/projects/matchbot/vendor/slim/slim/Slim/Middleware/RoutingMiddleware.php:59
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:147
/home/barney/projects/matchbot/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/home/barney/projects/matchbot/vendor/slim/slim/Slim/App.php:215
/home/barney/projects/matchbot/tests/Application/Actions/Donations/CreateTest.php:283
.

FAILURES!
Tests: 175, Assertions: 1015, Failures: 1.
barney@barney-l-xps:~/projects/matchbot$ docker compose exec app composer test
> XDEBUG_MODE=coverage phpunit
PHPUnit 9.5.21 #StandWithUkraine

[16-Dec-2022 14:33:56 UTC] PHP Deprecated:  Return type of LosMiddleware\RateLimit\RateLimitOptions::offsetGet($index) should either be compatible with ArrayObject::offsetGet(mixed $key): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/html/vendor/los/los-rate-limit/src/RateLimitOptions.php on line 46
......F........................................................  63 / 175 ( 36%)
............................................................... 126 / 175 ( 72%)
.................................................               175 / 175 (100%)

Time: 00:04.747, Memory: 30.00 MB

There was 1 failure:

1) MatchBot\Tests\Application\Actions\Donations\CreateTest::testNoXForwardedForHeader
Failed asserting that exception of type "Error" matches expected exception "LosMiddleware\RateLimit\Exception\MissingRequirement". Message was: "Call to a member function isSuccess() on null" at
/var/www/html/src/Application/Auth/DonationRecaptchaMiddleware.php:62
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:209
/var/www/html/vendor/middlewares/client-ip/src/ClientIp.php:65
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:147
/var/www/html/vendor/los/los-rate-limit/src/RateLimitMiddleware.php:59
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:209
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/var/www/html/vendor/slim/slim/Slim/Routing/Route.php:341
/var/www/html/vendor/slim/slim/Slim/Routing/RouteRunner.php:84
/var/www/html/app/routes.php:63
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:313
/var/www/html/vendor/slim/slim/Slim/Middleware/RoutingMiddleware.php:59
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:147
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/var/www/html/vendor/slim/slim/Slim/App.php:215
/var/www/html/tests/Application/Actions/Donations/CreateTest.php:283
.

FAILURES!
Tests: 175, Assertions: 1015, Failures: 1.
```

Mark test skipped throws an exception at that line and the test results show the test as skipped instead of failed. The `! str_contains(__DIR__, 'var')` is a way of checking that we're not in the docker container - so the test is only skipped on my local, it's not skipped on CircleCI where it's passing. This is what the output looks like with it skipped:

```
barney@barney-l-xps:~/projects/matchbot$ vendor/bin/phpunit 
PHPUnit 9.5.21 #StandWithUkraine

Warning:       No code coverage driver available

......S........................................................  63 / 175 ( 36%)
............................................................... 126 / 175 ( 72%)
.................................................               175 / 175 (100%)

Time: 00:03.423, Memory: 32.00 MB

OK, but incomplete, skipped, or risky tests!
Tests: 175, Assertions: 1014, Skipped: 1.
```

Reply:
@alihejazi https://github.com/thebiggive/matchbot/pull/464/files#r1050841897 probably makes it a bit clearer.

Reply:
@bdsl Thanks for the detail. I might be misunderstanding, but I don't understand why we're now skipping a test if we're running the code from outside of the container. I think this for two reasons:

1. The tests should be run from within docker anyways
2. If the test fails locally, but passes on CircleCI, then there's likely to be an issue with your local set up, or you've made a breaking change.

Perhaps get Noel's thought on this. You know much more about php than I do, but this seems a tad weird, unless I've misunderstood.

Reply:
It's not about a breaking change, or about inside or outside docker - the test fails in the same way on the main branch, running inside docker. See:

```
barney@barney-l-xps:~/projects/matchbot$ git show --stat; git status; docker compose exec app composer test
commit b58464580276ccdce426073296079ae478094a36 (HEAD -> main, origin/main)
Merge: b8a8921 96a888e
Author: Noel Light-Hilary <noel@thebiggive.org.uk>
Date:   Fri Dec 16 12:57:49 2022 +0000

    Merge pull request #466 from thebiggive/develop
    
    Develop -> main

 src/Domain/DonationRepository.php        |  2 +-
 src/Migrations/Version20221213144526.php | 30 ++++++++++++++++++++++++++++++
 src/Migrations/Version20221214100808.php | 31 +++++++++++++++++++++++++++++++
 3 files changed, 62 insertions(+), 1 deletion(-)
On branch main
Your branch is up-to-date with 'origin/main'.

nothing to commit, working tree clean
> XDEBUG_MODE=coverage phpunit
PHPUnit 9.5.21 #StandWithUkraine

[16-Dec-2022 14:58:08 UTC] PHP Deprecated:  Return type of LosMiddleware\RateLimit\RateLimitOptions::offsetGet($index) should either be compatible with ArrayObject::offsetGet(mixed $key): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/html/vendor/los/los-rate-limit/src/RateLimitOptions.php on line 46
......F........................................................  63 / 174 ( 36%)
............................................................... 126 / 174 ( 72%)
................................................                174 / 174 (100%)

Time: 00:04.688, Memory: 30.00 MB

There was 1 failure:

1) MatchBot\Tests\Application\Actions\Donations\CreateTest::testNoXForwardedForHeader
Failed asserting that exception of type "Error" matches expected exception "LosMiddleware\RateLimit\Exception\MissingRequirement". Message was: "Call to a member function isSuccess() on null" at
/var/www/html/src/Application/Auth/DonationRecaptchaMiddleware.php:62
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:209
/var/www/html/vendor/middlewares/client-ip/src/ClientIp.php:65
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:147
/var/www/html/vendor/los/los-rate-limit/src/RateLimitMiddleware.php:59
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:209
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/var/www/html/vendor/slim/slim/Slim/Routing/Route.php:341
/var/www/html/vendor/slim/slim/Slim/Routing/RouteRunner.php:84
/var/www/html/app/routes.php:63
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:313
/var/www/html/vendor/slim/slim/Slim/Middleware/RoutingMiddleware.php:59
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:147
/var/www/html/vendor/slim/slim/Slim/MiddlewareDispatcher.php:81
/var/www/html/vendor/slim/slim/Slim/App.php:215
/var/www/html/tests/Application/Actions/Donations/CreateTest.php:279
.

FAILURES!
Tests: 174, Assertions: 1011, Failures: 1.

Generating code coverage report in Clover XML format ... done [00:00.063]
Script XDEBUG_MODE=coverage phpunit handling the test event returned with error code 1
```

Reply:
@alihejazi I've deleted these new lines, I found I was able to fix my local by deleting the vendor folder and running composer install again. Apologies.